_includes/nav-sidebar.html

```diff
@@ -48,21 +48,71 @@
 
     {% assign url = "/projects/" | append: new_project[0].name | append: '/' %}
         <li class="{% if page_url[2] == new_project[0].name %}active{% endif %}">
-            <a class="" href="{{url}}">{{new_project[0].display_name}}</a>
+            <a class="" href="../{{url}}/tutorials/">{{new_project[0].display_name}}</a>
         <li>
   {% endfor %}
     </ul>
 {% endif %}
 
-
-{% if page_url[1] == "Tutorials" %}
-    <ul class="nav nav-sidebar">
-  {% assign tutorials = site.pages | where_exp: 'page', 'page.tutorial' | sort: 'tutorial-order' %}
-  {% for t in tutorials %}
-        <li class="{% if page.url == t.url %}active{% endif %}">
-            <a class="" href="{{t.url}}">{{t.title}}</a>
-        <li>
-  {% endfor %}
```

Comment:
I will need your help with this. The code deleted above was for ordering the tutorials on the old (current) docsite, which was pretty shoddy. I.e. the tutorial ordering made little sense, especially to the newbie.

Reply:
The Jekyll way of doing this would be to make the tutorials a collection, add an ordering property to the YAML front matter and add a [`sort_by`](https://jekyllrb.com/docs/collections/#sort-by-front-matter-key) to the config file to order it by that.

But this only works if all items are files in the repo. If that's not the case, things get more messy and the best solution is less clear. Another downside is that there is no overview of the order and you need to see the rendered result.

For the website I added [menu.json](https://github.com/seL4/website/blob/master/_data/menu.json) to have all metadata and ordering information in one place and used that as input to [generate](https://github.com/seL4/website/blob/master/_includes/menu.html) the wanted structure. The main reason I did it that way is to keep all the data compact with submenu support and in one place. It's also a mix of local and remote urls, so I couldn't add the metadata to front matter and I really didn't want to create shim files just for that as that's just messy.

Frankly, docs is a mess: It has all the tutorial shim files in the Tutorials directory which already contain a `tutorial-order` property, but it also added `_data/projects/sel4-tutorials.yml`. You only need one place to keep metadata.

What was wrong with the old code other than the order? The order can be changed by changing `tutorial-order`, it's probably alphabetically sorted by value. Main thing missing is the grouping you added, but that can be solved by adding a new field `category` to the first tutorial of each group: If set, generate a new '`<li>`'.

Once you know the order, it should also be possible to automatically generate the `Next tutorial` links too. Maintaining that manually will be hell and easy to forget when changing the order or adding new tutorials in the future.

You want to keep metadata out of your template code and you don't want the same metadata in multiple places.

Reply:
There are additional menu items in the tutorial side bar that are a mix of external links and other pages on the site. It also has non-link section headers. The nicest way would probably be similar to what you did for the main site, with one separate json file that has the structure and a separate `tutorials-nav` include.

Reply:
Hm. Nothing is ever simple. The `sel4-tutorials.yml` is currently there to define the "project", which has info that gets rendered on a bunch of overview pages, so we can't just remove it. We *can* use it to define the order, because it has an ordered list of all tutorial pages, and we can add additional info to each of the entries, but I'm not sure how to add additional info like section headers, external links, and links to other pages.

I guess, we could add another field "nav-section" to each component in there and then another top-level entry with a list of section headers, where each either has a nav section name or directly a list of nav items consisting of text and link. That could work. Not entirely sure I'll have time this afternoon, but I could give this a go.

Reply:
I would go for the minimal change: Keep the non-tutorial links hard coded here for now and use the existing `tutorial-order` to generate the tutorial part, with a new property to create submenus. And then you can use that to figure out the next tutorial link too.

Reply:
I couldn't get that quite working -- a lot of the hard coded part also needs to figure out "next" links, which is then not available as data. I went with something similar to the above, which removes a bit of the duplication (order and description in front matter) with sel4-tutorials.yml. The side bar works out ok, computing a reasonable "next" link did not end up to be very elegant.

Choosing the dict keys slightly differently would remove some of the case distinctions, but the ones prescribed by `sel4-tutorials.yml` sounded too nonsensical in the nav bar (`name` refers to file name, which I found most confusing in there. I guess `display_name` would be usable). The names in `sel4-tutorials.yml` is prescribed by other generic code on the site, so we can't tweak that without larger changes.

Anyway, have a look if the current version is acceptable. If not, we can always revert some of this.

I have not yet removed the manual "next" links for comparison. One thing that this brought up is that we have two "Overview" pages, which we should probably avoid.